message_edit: Show typing indicator for message editing.

roanster007 opened the pull request:
Previously, when a message was being edited, typing notification for it is not shown.

This commit adds typing indicators for message editing in stream as well as in Direct Messages, provided the send typing notification for corresponding is enabled.

Fixes #25719

<!-- Describe your pull request here.-->
API Design thread: https://chat.zulip.org/#narrow/stream/378-api-design/topic/typing.20notifications.20for.20editing.20messages/near/1720430


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**


![brave_aio5ftvP3Y](https://github.com/zulip/zulip/assets/97049524/3d98118c-792a-4735-a3fb-0a256928f369)
![brave_myobIZy0C9](https://github.com/zulip/zulip/assets/97049524/d3d15d21-877c-4497-9aba-6e2625c9caa4)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/views/typing.py:
```diff
@@ -12,22 +17,21 @@
-    req_type: str = REQ(
-        "type", str_validator=check_string_in(VALID_RECIPIENT_TYPES), default="direct"
-    ),
+    req_type: str = REQ("type", str_validator=check_string_in(VALID_RECIPIENT_TYPES)),
```
What's the thinking around removing the `default` property? I'm pretty sure that's required for existing mobile/terminal app versions that predate stream typing notifications.

roanster007 replied:
Previously, when typing notifications were available only to stream message edits and direct message edits, when the POST doesn't include `type="stream"`, it is taken as `type="direct"` by default. 

Now that we are adding extra types for message edit in these POST, I felt it would be better if the POST of direct message edits rather explicitly contained the `type="direct"` property, rather than `default="direct"`

timabbott replied:
So the intent is that any modern client should always pass this parameter. But clients that predate stream typing notifications and thus don't pass this parameter exist, including actually the current version of the mobile apps, and they need the default of "direct" to function.

You can add a comment explaining this to help others not "simplify" by removing it -- but it's also hinted in the API documentation for the endpoint, which is always worth reading when changing the API.

timabbott commented on zerver/views/typing.py:
```diff
@@ -45,6 +49,35 @@ def send_notification_backend(
+
+        stream = access_stream_by_id(user_profile, stream_id)[0]
+        access_stream_for_send_message(user_profile, stream, forwarder_user_profile=None)
+        do_send_message_edit_typing_notification(user_profile, stream_id, message_id, operator)
```
Why is a stream ID required for the message_edit case? I would think we could just have the client sent the message ID itself, and the server can figure out whether it's a stream or direct message, and just look up the message itself (and the user's access to and sender status of it) and simplify this code path greatly.

roanster007 replied:
I added stream ID so as to send the typing notification event only to the users subscribed to the stream in which the message is present. (I couldn't find any method that obtains the `stream_id` from the `message_id`)

roanster007 replied:
Updated!

timabbott commented on zerver/views/typing.py:
```diff
@@ -12,22 +17,21 @@
     ),
     stream_id: Optional[int] = REQ(json_validator=check_int, default=None),
     topic: Optional[str] = REQ("topic", default=None),
+    message_id: Optional[int] = REQ(json_validator=check_int, default=None),
```
I started an API design thread for some further details of this API design worth discussing as a group.

https://chat.zulip.org/#narrow/stream/378-api-design/topic/typing.20notifications.20for.20editing.20messages/near/1720430

roanster007 replied:
added it to PR description.

timabbott commented on web/styles/zulip.css:
```diff
@@ -3606,3 +3606,49 @@ select.invite-as {
+
+.dot:nth-child(3) {
+    animation: typing 1s infinite 0.6s;
+}
```
These CSS rules belong in `message_row.css`, since that's where we define CSS for the internals of messages. 

You should also give a less generic name than `dot`, and use SCSS scoping to nest the `nth-child` rules inside the base rule, etc. Look at how other message row CSS rules are defined.

roanster007 replied:
updated!

timabbott commented on web/styles/message_row.css:
```diff
@@ -178,6 +178,10 @@ $time_column_min_width: 42px; /* + padding */
 
+        .edit_notifications {
+            grid-area: edited;
+        }
```
This is so clean with CSS grid :)

timabbott commented on web/src/message_list_view.js:
```diff
@@ -356,6 +356,9 @@ export class MessageListView {
         const edit_history_details = analyze_edit_history(message_container.msg, last_edit_timestr);
+        message_container.editing_in_left_col = !include_sender && !is_hidden;
+        message_container.editing_alongside_sender = include_sender && !status_message;
+        message_container.editing_status_msg = include_sender && status_message;
```
How do these values differ from `edited_in_left_col` and friends?

roanster007 replied:
their values are same, but they differ from the `edited_in_left_col` in when their values are set. Like for `edited_in_left_col`, value initialized only after an edit is made, while for `editing_in_left_col`, I made a small modification to set its value every time an individual message is rendered.

timabbott replied:
OK. This is probably at least worth a comment ... but it might be better to just do a preparatory refactor to define properties like `message_edit_notices_in_left_col` for location logic, and then have a single variable for the edited content... Want to try doing that in a little preparatory PR that we can merge early?

roanster007 replied:
resolved in latest update!

timabbott commented on web/src/typing_events.js:
```diff
@@ -114,7 +139,12 @@ export function display_notification(event) {
+    }
 
     typing_data.kickstart_inbound_timer(
         key,
```
I'm not sure where in this logic it would be the case that when you navigate to a message that someone is actively editing, that fact would be visible.

roanster007 replied:
The way that this currently works is -- the typing notification event for a message is sent to all the users subscribed to the stream in which the message is present. So irrespective of whether the user is viewing the message or not, the event would add the `. . .` animation  to the `editing_notifications` <div> present.

However, when a user subscribes to a stream just after a message editing in that stream has started, then he may have to wait until next push for the notification.

roanster007 replied:
Ok, I get it. I have updated the code to maintain the states of `message_id` being edited in `typing_data`, which is used by `message_list_view` to render the message edit animation, when user navigates to the stream after event is received.

timabbott replied:
OK and how does the message get its `message_container.is_typing` cleared when typing stops? I'm somewhat worried that a background rerender might cause incorrect display, but maybe the theory is that `set_calculated_message_container_variables` gets called every time a message is rerendered? If so this is likely correct.

roanster007 replied:
yes, `set_calculated_message_container_variables` is called every re render.

roanster007 replied:
This is resolved in the latest update.

timabbott commented:
Thanks for working on this @roanster007! I posted a batch of comments that you can get started on while we finalize the API design for this.

roanster007 commented on web/src/typing_events.js:
```diff
@@ -77,6 +79,21 @@ export function render_notifications_for_narrow() {
+    if (is_typing) {
+        $row.find(".message_edit_notice").hide();
+    } else {
+        $row.find(".message_edit_notice").hide();
```
Oh sorry! might have missed that while adjusting with the tests... I'll update it.

roanster007 commented on web/src/message_list_view.js:
```diff
@@ -780,6 +787,17 @@ export class MessageListView {
+        const status_message = Boolean(message_container.status_message);
+        message_container.editing_in_left_col = !include_sender && !is_hidden;
+        message_container.editing_alongside_sender = include_sender && !status_message;
+        message_container.editing_status_msg = include_sender && status_message;
```
When a new message is sent to a stream and is edited such that edit happens in the left_col , then for the clients the typing notification for editing won't appear. Hence, this section is added when only a single message is rendered.

timabbott replied:
I'm pretty sure the block in https://github.com/zulip/zulip/pull/28585/files#r1453938697 is expected to run every time a message is being rendered; and it looks like it only inspects immutable properties of the message, so I'd expect it to compute the same value every time. I would try deleting this and then debugging with `console.log` or breakpoints or something why that isn't working for you. From an abstractions standpoint, the other code path is where this belongs.

roanster007 commented:
@timabbott , I have updated the PR as per the comments, except for that preparatory PR. I explained the difficulty I am facing in that as a reply there.

alya commented:
Testing the PR, the dots are invisible in light theme.

roanster007 commented:
Ah, sorry for the oversight! @alya , I have updated the PR, screencap. could you confirm if the dots are okay for dark mode or they need to be brighter. (they are now of same colour as the `User is typing...` notification)

alya commented:
It looks reasonable to me; we can let folks comment further when it's deployed on CZO.

alya commented:
@timabbott I think this is ready for you to take another look.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,12 @@ format used by the Zulip server that they are interacting with.
+
+* [`POST /message_edit_typing`](/api/set-message-edit-typing-status):
+  Added support for sending typing notification when a message is
+  being edited both in streams and direct messages.
```
This should also document the new event type in `GET /events`.

roanster007 replied:
I updated it.

timabbott commented on web/src/message_edit.js:
```diff
@@ -348,6 +348,7 @@ function handle_message_row_edit_keydown(e) {
                     return;
                 }
                 save_message_row_edit($row);
+                $(document).trigger("message_editing_done.zulip");
```
I think it's likely better for this sort of trigger to be inside `save_message_row_edit` -- it's a bad pattern to split logic between the click handler and the function that actually does stuff -- and in particular will create bugs where this event is sent incorrectly if `save_message_row_edit` fails without throwing an exception.

timabbott commented on web/src/typing.js:
```diff
@@ -56,6 +75,19 @@ function send_typing_notification_based_on_message_type(to, operation) {
 }
 
+function send_typing_notifications_for_message_edit(message_id, operation) {
+    if (get_message_edit_typing_notification_policy()) {
```
This function is poorly named; the name in no way indicates the type is boolean rather than an enum that is the value of the policy, which one would not want to use in this sort of conditional. Call it something like `message_edit_typing_notifications_enabled()`.

timabbott replied:
Oh and it should accept the message object or ID, and check if that message is a stream/direct message to determine which policy to apply. Looking at `narrow_state` is incorrect -- views like "all messages" can contain both direct and stream messages, and we need to check the policy for the actual message in question.

roanster007 replied:
I guess in that case, `message_store` should do the work right...?

timabbott commented on web/src/typing.js:
```diff
@@ -79,11 +111,19 @@ function get_current_time() {
+        send_typing_notifications_for_message_edit(to, "start");
+    } else {
+        send_typing_notification_based_on_message_type(to, "start");
+    }
```
We're not going to merge a version which uses these hackery `number` types to determine what type of object this is. Let's get this file migrated to `TypeScript` and use actual types for our local data structures.

timabbott commented on web/src/typing.js:
```diff
@@ -129,4 +181,22 @@ export function initialize() {
+            page_params.server_typing_started_wait_period_milliseconds,
+            page_params.server_typing_stopped_wait_period_milliseconds,
+        );
+    });
```
Please don't register extra click handlers like this. This should be done via the `message_edit` module handlers for the save/cancel operations, so that these calls can be thoughtfully sequenced with the rest of that code path and its error handling code. (E.g. if a click is not actually processed by that code, it should not issue these events).

roanster007 replied:
The reason I registered this click handler here is the `worker` object which is needed for the call back of the click event.

timabbott commented on web/src/typing_data.ts:
```diff
@@ -64,6 +65,25 @@ export function get_topic_typists(stream_id: number, topic: string): number[] {
+    if (!edit_message_typing_ids.includes(message_id)) {
+        edit_message_typing_ids.push(message_id);
+    }
+}
```
Don't use a list data structure, please use a Map or a Set. Efficiency doesn't matter here as one is unlikely to have dozens of these, but one should always use correct data structure types

timabbott commented on web/src/typing_data.ts:
```diff
@@ -64,6 +65,25 @@ export function get_topic_typists(stream_id: number, topic: string): number[] {
+    }
+    const message_id_index = edit_message_typing_ids.indexOf(message_id);
+    edit_message_typing_ids.splice(message_id_index, 1);
+    return true;
```
Whenever you find yourself writing messy splicing logic like this, you should stop yourself as it's a sign you've got the wrong data structure.

roanster007 replied:
ok right... I'll use a Set in that case.

timabbott commented on web/src/typing_events.js:
```diff
@@ -77,6 +79,22 @@ export function render_notifications_for_narrow() {
+            typing: is_typing,
+        }),
+    );
+}
```
Does this logic correctly handle multiple rendered message lists? You can likely test using the `s` keyboard shortcut to flip between "All messages" (which is always rendered) and a narrowed view.

roanster007 replied:
I tested it manually.... it does handle the "All messages" list and narrowed ones (to both stream and topic) correctly.

roanster007 replied:
yup, you're right. This logic works as far as the container is re rendered when you change narrows/views. However, for "All Messages", since it is already present and we don't re render anything.... I'll try to think of something else.

timabbott commented on web/templates/message_body.hbs:
```diff
@@ -55,8 +73,18 @@
 {{#if edited_in_left_col}}
+{{#unless is_typing}}
 {{> edited_notice}}
+{{/unless}}
```
These changes are super messy. We should do a preparatory PR that refactors to have the template render with unconditionally set variables `edited_notices_in_left_col` and then just have an `is_typing` if/else block added inside that container.

You can do that preparatory refactoring PR and then come back to this -- we should be able to merge that change on top of `main`.

roanster007 replied:
Updated!

timabbott commented on zerver/views/typing.py:
```diff
@@ -60,3 +67,42 @@ def send_notification_backend(
+    message_id: Optional[int] = REQ(json_validator=check_int, default=None),
+) -> HttpResponse:
+    if message_id is None:
+        raise JsonableError(_("Missing message_id"))
```
Just remove the `default=None` parameter that makes this parameter not required, and you can delete this check -- the `REQ` framework is designed to take care of this sort of thing for you.

roanster007 replied:
updated.

timabbott commented on zerver/views/typing.py:
```diff
@@ -60,3 +67,42 @@ def send_notification_backend(
+    recipient = message.recipient
+
+    if message.sender_id != user_profile.id:
+        raise JsonableError(_("You don't have permission to edit this message"))
```
This permissions check is not correct. Message editing can be disabled in the organization, or time-limited. You need to look at the logic we use in `message_edit.py` to check, and ideally call the same core function it uses to check permissions.

timabbott commented on zerver/views/typing.py:
```diff
@@ -60,3 +67,42 @@ def send_notification_backend(
+            raise JsonableError(_("User has disabled typing notifications for stream messages"))
+
+        stream_id = recipient.type_id
+        do_send_message_edit_typing_notification(user_profile, stream_id, message_id, operator)
```
`do_send_stream_message_edit_typing_notification`

roanster007 commented on web/shared/src/typing_status.ts:
```diff
@@ -12,7 +12,11 @@ export type Recipient =
-      });
+      })
+    | {
+          message_type: "single_message";
```
`message_type` = `single_message` might be a weird name for this.... I thought of it since typing notifications for edit messages are more like dealing with individual messages than a `stream` or `direct` type.

timabbott replied:
I think probably you should just add a separate field, maybe called `notification_event_type` field that can be `typing` or `typing_message_edit` (or whatever we picked in API design).

It's best to use the same terms across the code base for the same idea, since that makes `git grep` nice, and so a new `single_message` type is just confusing.

roanster007 commented on web/templates/edited_notice.hbs:
```diff
@@ -1,4 +1,9 @@
+<div id="editing_notifications" class="edit_notifications"></div>
+{{#if is_typing}}
+    {{> editing_notifications typing=true}}
+{{/if}}
```
In the above code, the `is_typing` is used to check if the message is being edited when you are rendering that particular message_container, and display the notification accordingly. Once the typing is done, this is hidden until you re render that container. 

The `id=editing_notifications` div is used for showing typing notifications, when you already have the `message_container` rendered (basically that message is within your view).

roanster007 commented:
I addressed all the reviews.

roanster007 commented on web/src/typing_events.ts:
```diff
@@ -104,6 +113,27 @@ export function render_notifications_for_narrow(): void {
+    if (is_typing) {
+        $row?.find(".message_edit_notice").hide();
+        $all_msgs_view_row?.find(".message_edit_notice").hide();
+    } else {
```
Unlike the other narrows wherein the `message_containers` are gonna be re rendered, since in "All Messages", they stay there all the time, for each message editing, along with displaying of notification in the current narrow, I am also adding/removing the notification in the "All Messages" view.

roanster007 commented on web/src/typing_events.ts:
```diff
@@ -104,6 +113,29 @@ export function render_notifications_for_narrow(): void {
+
+    const $all_msgs_view_row = message_lists.home?.get_row(message_id);
+    apply_message_edit_notifications($all_msgs_view_row, is_typing);
+}
```
So since "All Messages" is gonna be staying all the time without re render every time you click on it, I am dealing it separately by adding/removing the typing notification each time.

roanster007 commented on web/templates/edited_notice.hbs:
```diff
@@ -1,4 +1,24 @@
+        <div class="message_edit_notice" data-tippy-content="{{t 'Last edited {last_edit_timestr}.'}}" style="display: none;">
+            {{t "EDITED"}}
+        </div>
+    {{/if}}
```
I made separate `display:none` and without `display:none` since like when you narrow to a stream when a message is being edited, you still want the `EDITED` notice in your DOM, but just hidden so that you can unhide it when the editing is done.

roanster007 commented:
@timabbott I have addressed all the reviews, and have fixed all the possible bugs which I figure out...Please do point out if you find something off regarding the logic.

roanster007 commented:
Removed the conflicts... zulipbot didn't update the status

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,18 @@ format used by the Zulip server that they are interacting with.
+  is sent when a user starts editing a message.
+  It sends fields like `op` indicating whether user has started editing or not,
+  `sender` representing the user who started editing the message, and `message_id`
+  indicating the id of the message that is being edited.
```
Delete this second paragraph, it's too in the details for this changelog.

timabbott commented on web/shared/src/typing_status.ts:
```diff
@@ -41,12 +45,20 @@ function same_recipient(a: Recipient | null, b: Recipient | null): boolean {
+        } else if (a.message_type === "stream" && b.message_type === "stream") {
+            // stream recipients
+            return same_stream_and_topic(a, b);
+        }
```
I'm not sure whether or not this logic makes sense. The previous code was reliant on the assumption that at most one message can be in the compose box at a time.

This new `typing_message_edit` event type involves a separate input box from the compose box, and so possibly should be treated as a parallel/independent input.

roanster007 replied:
So in case of `message editing` this method basically checks out if the new notification received, is it of the same recipient as that of the current one. If yes, then continue showing the editing notification.

So in case when you are editing one message, and before the notification of this expires, move to another message and start editing that, then by the current method, the notification of the previous message editing would be stopped, and the editing notification of current message would be shown.

Do you want the previous message editing notification to be still displayed along with the current message's notification until it expires naturally?

timabbott replied:
I talked with @alya and she confirmed that the product model that makes sense for this is to just have typing notifications tracked separately for different inputs.

(It's certainly possible to have multiple compose/edit inputs open and with current work in them at the same time).

I think if you refactor to support that model, it'll be a lot easier to read/debug/verify than doing this overloading.

roanster007 replied:
I'll update it.

timabbott commented on web/src/message_edit.js:
```diff
@@ -1044,13 +1045,15 @@ export function save_message_row_edit($row) {
         url: "/json/messages/" + message.id,
         data: request,
         success() {
+            $(document).trigger("message_editing_done.zulip");
```
What's the thinking around using jQuery events for this? In general we have not added any new ones in years because they have unpredictable behavior and we plan to deprecate jQuery itself eventually.

It's also just awkward that this doesn't pass the `ID` of the message being edited as a parameter, which is clearly important for handlers to correct manage this. Why not just have `message_edit.js` call a function in the `typing*` modules directly?

roanster007 replied:
Regarding the passing of `message_id` as a parameter -- I guess if we want editing notification of only one single message at a time (if while editing a message, you move on to editing another message, then the edit notification of first message should be cancelled.), then I guess the `message_id` wouldn't be needed inside the method.

roanster007 replied:
updated!

timabbott commented on web/src/typing.ts:
```diff
@@ -71,13 +92,33 @@ function send_typing_notification_based_on_message_type(
 ): void {
+    if (to.notification_event_type !== "typing") {
+        return;
+    }
```
Should this just be an `assert to.notification_event_type === "typing"`? 

Writing this logic with a `return` statement makes one have to think about whether returning is correct that case (and what that case is), so if it's intended to be impossible, we should just say that.

roanster007 replied:
thanks for letting me know! updated it.

timabbott commented on web/src/typing.ts:
```diff
@@ -71,13 +92,33 @@ function send_typing_notification_based_on_message_type(
+
+function message_edit_typing_notifications_enabled(message_id: number): boolean {
+    const message = message_lists.current?.get(message_id);
+    if (message?.type === "stream") {
```
Should you be using `assert` statements or `!` here? I think if you look up the message in `message_store` rather than `message_lists.current`, we should be able to avoid thinking about whether `current` is undefined. And that should only fail on invalid message IDs or those that have been deleted from the message store, and I would hope that this code path would not run for a deleted message. (You might need some code to do cleanup in the "delete message" handler code paths).

And certainly one should be able to assume that a message object has a `type`  field.

roanster007 commented:
had a really tough time trying to put up the refactoring commit into this as the first commit due to some merge conflicts.... So I rather made the same refactoring in the pr #29066

roanster007 commented:
@timabbott could you have another look at it?

roanster007 commented:
addressed all reviews! Ready for another review.

whilstsomebody commented:
@roanster007 , can I continue with your work on this?

timabbott commented:
Closing in favor of #33024, which continues this effort.
